### PR TITLE
[extension/ballast] Remove the deprecated memory ballast extension

### PR DESCRIPTION
**Pull request by TylerHelmuth**
#### Description
This PR removes the deprecated memory ballast extension and all the logic in place in memorylimiter and service that was using it.

#### Link to tracking issue
Related to https://github.com/open-telemetry/opentelemetry-collector/issues/8343. I don't want to close it until the [helm chart](https://github.com/open-telemetry/opentelemetry-helm-charts/pull/1268) is updated.

#### Testing
Unit tests

### CODE REVIEW

File: .chloggen/remove-ballast-extension.yaml
Comment (crobert-1) [inline .chloggen/remove-ballast-extension.yaml:13]:
```suggestion
issues: [8343]
```
Nit: I assume we should reference the original issue here?

File: cmd/otelcorecol/go.mod
Comment (crobert-1) [inline cmd/otelcorecol/go.mod:7]:
```suggestion
toolchain go1.21.12
```
I assume this is unrelated to this PR and we should remove this change?

Comment (atoulme) [reply cmd/otelcorecol/go.mod:7]: yes that's a good assumption

File: .chloggen/remove-ballast-extension.yaml
Comment (TylerHelmuth) [reply .chloggen/remove-ballast-extension.yaml:13]: Our practice is to use the PR so that the changelog links to the exact commit that did the thing.  he description links to the issue so you can quickly get to it.

**Review by mx-psi — COMMENTED** [review]
Can we possibly split this process in more steps?

For example, a first step before outright removal can be:

- Removing it from the Helm chart
- Removing support from it in the service (that is, fixing #8342) and documenting this
- Removing any mention of it on docs and examples on the opentelemetry org

Then a second step can be removing it from the official binaries while keeping a source-only release, and a final step would be removing the source code entirely.

I think this would solve our immediate issue and make the deprecation process more manageable (not all users would get this removed at the same time, so support requests are more easy to manage).

I know this is more annoying but 
- this extension has been here since forever
- we don't know which users have been listening to our deprecations

so I think we need to be gentle and divide this into steps

**Review by atoulme — APPROVED** [review]

**Review by songy23 — APPROVED** [review]

**Review by mx-psi — COMMENTED** [review]
Have we seen any users complaining after open-telemetry/opentelemetry-collector-releases/pull/607? I see #10995, have we had any other reports?

If #10995 is the only one, I am comfortable merging this

**Review by mx-psi — APPROVED** [review]
Oki, then I would propose we wait until, say, Wednesday next week and merge if we don't get more reports

**Review by codeboten — APPROVED** [review]

**Pull request merged**
